dash-renderer/src/observers/prioritizedCallbacks.ts

```diff
@@ -0,0 +1,104 @@
+import {
+    flatten,
+    map,
+    reduce,
+    uniq,
+    pluck,
+    partition,
+    includes
+} from 'ramda';
+
+import { IStoreState } from "../store";
+
+import {
+    aggregateCallbacks,
+    removePrioritizedCallbacks,
+    addExecutingCallbacks,
+    executeCallback
+} from '../actions/callbacks';
+
+import { stringifyId } from '../actions/dependencies';
+
+import {
+    combineIdAndProp
+} from '../actions/dependencies_ts';
+
+import isAppReady from '../actions/isAppReady';
+
+import {
+    ICallback,
+    IExecutingCallback,
+    ICallbackProperty
+} from '../types/callbacks';
+import { IStoreObserverDefinition } from '../StoreObserver';
+
+const observer: IStoreObserverDefinition<IStoreState> = {
+    observer: async ({
+        dispatch,
+        getState
+    }) => {
+        const { callbacks: { executing, watched }, config, hooks, layout, paths } = getState();
+        let { callbacks: { prioritized } } = getState();
+
+        const available = Math.max(
+            0,
+            10 - executing.length - watched.length
```

Comment:
Most browsers don't allow more than 6 concurrent connections to the same domain. By over-provisioning, attempting to grossly balance useless callback execution (eg. subsequently dropped) vs. suffering from callback starvation.

Reply:
> subsequently dropped

the case where callbacks are returned out of order and then their responses ignored?

Reply:
I’d imagine that it would be faster for the browser to manage queuing and dispatching the extra requests would be faster than us managing and dispatching the requests, since there wouldn’t be any blocking JS in between request dispatches.

But I suppose the trade off is how many of these responses end up getting dropped on our end. So, when do sponsors get dropped? 

One case would be when there are multiple requests to the same callback, like typing data into a dcc.Input or dragging a slider. In that case, I suppose we would take the time to introspect the pending callbacks to see if there are any redundant updates and, if so, skip the requests that were triggered earlier? eg if a slider is dragged from 1 to 20, then we might make 5 requests with values from 1-5, and then before we start the next batch, we’d see that the remaining 15 pending requests are to the same callback so we’d drop requests 5-19 and only dispatch the request for 20. 

Am I thinking about this the right way? Are there other use cases where this comes into play besides multiple updates to the same component?

Reply:
> I’d imagine that it would be faster for the browser to manage queuing and dispatching the extra requests would be faster than us managing and dispatching the requests, since there wouldn’t be any blocking JS in between request dispatches.

Technically this is true but under certain scenarios, keep the flexibility to prune out the callbacks before the request is made can result in better real-world performance, with the caveat that we need to be careful about preventing sub-optimal usage of concurrent connections.

For example, when loading dash-docs `/all` with the `dev` version of the renderer, at some point ~155 callbacks will be "pending" - that is assigned for execution but blocked at the browser level from executing because of the concurrency rules.

> So, when do responses get dropped?

Callbacks can be pruned at any state up to when execution completes. Pruning prior to execution is by far preferable if possible as it doesn't use up available connections. Specifically pruning is done on each callback completion and request (https://github.com/plotly/dash/pull/1254/files#diff-178b2bd05a773877fd1b117eef8b1e8cR124) for all non-completed callbacks.

Let's imagine that the user clicks a link to go from dash-docs `/all` to `/something-else` - the `dccLocation.pathname` callback *needs* to wait for 150 of the 155 callbacks to complete before it's gets executed as they are already in the queue and out of our control.

While this is not currently part of the PR, doing something as simple as prioritizing callbacks that were triggered by user interactions (this should be a relatively simple addition), over component derived props, over layout change chains would make it possible to execute the user interaction callback first and drop the large majority of the 155 callbacks by layout/pruning, as most would not have been assigned already.

It is also possible to `abort` XHR requests once they've been requested but this is problematic for the scenario above. The callback's result is what causes the 155 previous callbacks to be invalidated, if these are already in queue, the new one can't take precedence.